client/trino-jdbc/src/main/java/io/trino/jdbc/TrinoResultSet.java

```diff
@@ -82,9 +91,39 @@ public QueryStats getStats()
     public void close()
             throws SQLException
     {
-        if (closed.compareAndSet(false, true)) {
-            ((AsyncIterator<?>) results).cancel();
-            client.close();
+        try (Closer closer = Closer.create()) {
+            synchronized (this) {
+                if (closed) {
+                    // Already closed
+                    return;
+                }
+                closed = true;
+
+                for (Runnable runnable : Lists.reverse(afterClose)) {
```

Comment:
`Closer` already runs in reverse order, so I assume the intent here is to close the statements in the order they were created. Maybe add a comment to that effect. Though we could close them in any order, so it'd be simpler to just add them in the default `Closer` order.

Reply:
Actually, I don't think we need a list at all. This `runAfterClose` feature is to close *the* statement for a result set. We never have more than one. So we could have a single field and simplify this.